File: csc246-fall2018-assignment1/square.c
/**
  * CSC246 Assignment 1 problem 3.
  *
  * @author John Choi
  * @version 09202018
  */
#include <unistd.h>
#include <stdio.h>
#include <ctype.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <stdbool.h>

// Size of the square we're looking for.
#define SQUARE_WIDTH 6
#define SQUARE_HEIGHT 6

// Print out an error message and exit.
static void fail(char const *message) {
  fprintf(stderr, "%s\n", message);
  exit(1);
}

// Print out a usage message, then exit.
static void usage() {
  printf("usage: square <workers>\n");
  printf("       square <workers> report\n");
  exit(1);
}

// Size of the grid of characters.
int rows, cols;

// Maximum width of a row.  Makes it easier to allocate the whole
// grid contiguously.
#define MAX_WIDTH 16384

// Type used for a row of the grid.  Makes it easier to declare the
// grid as a pointer.
typedef char Row[MAX_WIDTH];

// Grid of letters.
Row *grid;

// Read the grid of characters.
void readGrid() {
  // Read grid dimensions.
  scanf("%d%d", &rows, &cols);
  if (cols > MAX_WIDTH) {
    fprintf(stderr, "Input grid is too wide.\n");
    exit(EXIT_FAILURE);
  }

  // Make space to store the grid as a big, contiguous array.
  grid = (Row *)malloc(rows * sizeof(Row));

  // Read each row of the grid as a string, then copy everything
  // but the null terminator into the grid array.
  char buffer[MAX_WIDTH + 1];
  for (int r = 0; r < rows; r++) {
    scanf("%s", buffer);
    memcpy(grid[r], buffer, cols);
  }
}

/* keeps in track of the current row */
int volatile currentRow = 0;
/* number of workers */
int numWorker;
/* number of current running worker */
int volatile currentWorker;

/**
  * Checks individual 6 by 6 squares to see if it is a valid square.
  *
  * @param squares - individual squares to check
  * @return true if this square is valid
  */
bool checkSquare(char squares[6][6]) { //TODO algorithm for checking to see if every alphabet exists
  char *knownLetters = (char *)malloc(26 * sizeof(char));
  for (int i = 0; i < 26; i++) {
    *(knownLetters + i) = '*';
  }
  int counter = 0;
  for (int i = 0; i < 6; i++) {
    for (int j = 0; j < 6; j++) {
      bool exists = false;
      for (int k = 0; k <= counter; k++) {
        if (squares[i][j] == *(knownLetters + k)) { //if letter is found
          exists = true;
        }
      }
      if (!exists) {
        *(knownLetters + counter) = squares[i][j];
        counter++;
      }
    }
  }
  free(knownLetters);
  if (counter != 26) {
    return false;
  }
  return true;
}

/* Global variable that holds true if user passed in report argument */
bool report;

/**
  * Searches and finds 6 x 6 square in the grid.
  *
  * @return number of squares found
  */
int findSquares() {
  int totalSquares = 0;
  char square[6][6];
  int lowerBoundRow = currentWorker + (currentRow * numWorker);
  int upperBoundRow = lowerBoundRow + 5;

  int lowerBoundCol = 0;
  int upperBoundCol = lowerBoundCol + 5;
  int currentLocationA = 0; //used for square
  int currentLocationB = 0; //used for square

  while (upperBoundRow < rows) { //one process increments - scans vertically
    while (upperBoundCol < cols) { //scans horizontally
      for (int i = lowerBoundRow; i <= upperBoundRow; i++) { //for row
        for (int j = lowerBoundCol; j <= upperBoundCol; j++) { // for col
          square[currentLocationA][currentLocationB++] = grid[i][j];
        }

        currentLocationA++;
        currentLocationB = 0;
      }
      currentLocationA = 0;

      if (checkSquare(square)) {
        if (report) {
          fprintf(stdout, "%d %d\n", lowerBoundRow, lowerBoundCol);
        }

        totalSquares++;
      }
      lowerBoundCol++;
      upperBoundCol++;
    }
    currentRow++;
    lowerBoundRow = currentWorker + (currentRow * numWorker);
    upperBoundRow = lowerBoundRow + 5;
    lowerBoundCol = 0;
    upperBoundCol = lowerBoundCol + 5;
  }
  return totalSquares;
}

/**
  * Entry point for this program.
  *
  * @param argc number of arguments
  * @param argv list of arguments
  * @return exit status
  */
int main(int argc, char *argv[]) {
  report = false;
  int workers = 4; //number of child processes to create

  // Parse command-line arguments.
  if (argc < 2 || argc > 3)
    usage();

  if (sscanf(argv[1], "%d", &workers) != 1 || workers < 1)
    usage();

  // If there's a second argument, it better be "report"
  if (argc == 3) {
    if (strcmp(argv[2], "report") != 0)
      usage();
    report = true;
  }

  readGrid();

  // ...
  numWorker = workers;
  int totalSquares = 0;
  int pfd[2]; //read and write pipe
  pid_t pids[workers];
  currentWorker = 0;
  for (int i = 0; i < workers; i++) {
    if (pipe(pfd) != 0) {
      fail("Can't create pipe");
    }

    if ((pids[i] = fork()) == -1) {
      fail("Cannot create child");
    } else if (pids[i] == 0) { //child
      close(pfd[0]); //close the reading end of the pipe
      int numOfSquares = findSquares();
      lockf(pfd[1], F_LOCK, 0);
      char sNum[6];
      sprintf(sNum, "%d", numOfSquares); //converts number of squares to string
      //send the number of squares to parent
      int num = write(pfd[1], sNum, 6);
      if (num < 0) {
        fail("Cannot write into pipe");
      }
      lockf(pfd[1], F_ULOCK, 0);
      close(pfd[1]);
      return EXIT_SUCCESS;
    } else { //parent
      close(pfd[1]); //close the writing end of the file
      //read from the pipe
      int numOfSquares;
      char sNum[6];
      if (read(pfd[0], sNum, 6) < 0) { //reads from the pipe
        fail("Error reading from the pipe");
      }
      numOfSquares = atoi(sNum); //converts string to int
      totalSquares += numOfSquares;
    }
    currentWorker++;
  }
  // wait for children to exit
  int status;
  while (workers > 0) {
    wait(&status);
    workers--;
  }
  char totalSquareString[6];
  sprintf(totalSquareString, "%d\n", totalSquares);
  write(STDOUT_FILENO, "Squares: ", 10);
  write(STDOUT_FILENO, totalSquareString, strlen(totalSquareString));
  return EXIT_SUCCESS;
}

File: csc246-fall2018-assignment2/Square.java
import java.util.ArrayList;
import java.util.Scanner;

/**
 * CSC246 Assignment 2 Problem 3.
 *
 * @author John Choi
 * @since 10052018
 */
public class Square {

	private static volatile int numSquares = 0;
	private static volatile boolean report;
	private static char[][] board;
	private static int rows;
	private static int cols;
	private static Scanner input = new Scanner(System.in);
	private static volatile int currentRow = 0;
	private static volatile int currentWorker;

	/**
	 * Inner class that handles thread operation.
	 *
	 * @author John Choi
	 * @since 10052018
	 */
	static class SingleThread implements Runnable {

		private int numWorker;


		public SingleThread(int numWorker) {
			this.numWorker = numWorker;
		}

		/**
		 * Defines the thread's job.
		 */
		@Override
		public void run() {
			findSquares();
		}

		/**
		 * Returns true if the passed in 2d array is a valid square.
		 *
		 * @param square to validate
		 * @return true if square is valid
		 */
		private boolean isSquare(char[][] square) {
			ArrayList<Character> alphabets = new ArrayList<>();
			for (char i = 'a'; i <= 'z'; i++) {
				alphabets.add(i);
			}

			for (int i = 0; i < square.length; i++) {
				for (int j = 0; j < square[0].length; j++) {
					if (alphabets.contains(square[i][j])) {
						int idx = -1;
						for (int k = 0; k < alphabets.size(); k++) {
							if (alphabets.get(k) == square[i][j]) {
								idx = k;
								break;
							}
						}
						alphabets.remove(idx);
					}
				}
			}
			if (alphabets.size() == 0) {
				return true;
			}
			return false;
		}

		/**
		 * Searches and finds 6 x 6 square in the board.
		 */
		public void findSquares() {
			int totalSquaresFound = 0;
			currentRow = 0;
			char[][] square = new char[6][6];
			int lowerBoundRow = currentWorker + (currentRow * numWorker);
			int upperBoundRow = lowerBoundRow + 5;
			int lowerBoundCol = 0;
			int upperBoundCol = lowerBoundCol + 5;
			int currentLocationA = 0; //used for square
			int currentLocationB = 0; //used for square
			while (upperBoundRow < rows) { //one process increments - scans vertically
				while (upperBoundCol < cols) { //scans horizontally
					for (int i = lowerBoundRow; i <= upperBoundRow; i++) { //for row
						for (int j = lowerBoundCol; j <= upperBoundCol; j++) { // for col
							square[currentLocationA][currentLocationB++] = board[i][j];
						}

						currentLocationA++;
						currentLocationB = 0;
					}
					currentLocationA = 0;

					if (isSquare(square)) {
						if (report) {
							System.out.printf("%d %d\n", lowerBoundRow, lowerBoundCol);
						}

						totalSquaresFound++;
					}
					lowerBoundCol++;
					upperBoundCol++;
				}
				currentRow++;
				lowerBoundRow = currentWorker + (currentRow * numWorker);
				upperBoundRow = lowerBoundRow + 5;
				lowerBoundCol = 0;
				upperBoundCol = lowerBoundCol + 5;
			}
			numSquares += totalSquaresFound;
		}
	}

	/**
	 * Initializes the board.
	 */
	public static void initializeBoard() {
		rows = input.nextInt();
		cols = input.nextInt();
		input.nextLine();
		board = new char[rows][cols];
		for (int i = 0; i < rows; i++) {
			String thisline = input.nextLine(); //save one line
			for (int j = 0; j < cols; j++) {
				board[i][j] = thisline.charAt(j);
			}
		}
		input.close();
	}

	/**
	 * Main method for this program.
	 *
	 * @param args command-line arguments
	 */
	public static void main(String[] args) {
		report = false;
		if (args.length == 2) {
			report = true;
		} else if (args.length == 1) {
			report = false;
		} else {
			System.out.println("usage: java Square <number-of-threads> < filename.txt");
			System.out.println("       java Square <number-of-threads> report < filename.txt");
			System.exit(1);
		}

		initializeBoard();
		int numOfThreads = Integer.parseInt(args[0]);
		Thread[] threads = new Thread[numOfThreads];
		currentWorker = 0;
		for (int i = 0; i < numOfThreads; i++) {
			threads[i] = new Thread(new SingleThread(numOfThreads));
			threads[i].run();
			currentWorker++;
		}

		for (int i = 0; i < numOfThreads; i++) {
			try {
				threads[i].join();
			} catch (InterruptedException e) {
				e.printStackTrace();
			}
		}
		System.out.println("Squares: " + numSquares);
	}
}
